File: travaux_diriges/tp1/hypercube.py
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()
token = None

if rank == 0:
    token = 42
    print(f"Process {rank} initializes the token with value {token}")

d = size.bit_length() - 1  # Dimension of the hypercube

# Diffusion of the token in the hypercube
for i in range(d):
    partner = rank ^ (1 << i)
    if rank < partner:
        if token is not None:
            comm.send(token, dest=partner)
    else:
        token = comm.recv(source=partner)
    comm.Barrier()

print(f"Process {rank} received token: {token}")

MPI.Finalize()

File: travaux_diriges/tp1/hypercube.cpp
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(int argc, char *argv[]) {
    int rank, size, dimension, token = 100; // 初始化 token 为 100

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    dimension = (int)log2(size);
    if (pow(2, dimension) != size) {
        if (rank == 0) {
            printf("Error: Number of processes must be a power of 2.\n");
        }
        MPI_Finalize();
        return EXIT_FAILURE;
    }

    if (rank == 0) {
        printf("Process %d initializes token = %d\n", rank, token);
    }

    for (int d = 0; d < dimension; d++) {
        int partner = rank ^ (1 << d);

        if (rank < partner) {
            MPI_Send(&token, 1, MPI_INT, partner, 0, MPI_COMM_WORLD);
        } else {
            MPI_Recv(&token, 1, MPI_INT, partner, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
        }
        printf("Process %d received token = %d from process %d at dimension %d\n", rank, token, partner, d);
    }

    MPI_Finalize();
    return EXIT_SUCCESS;
}
